src/Database/LSMTree/Internal/Paths.hs

```diff
@@ -16,33 +24,100 @@ module Database.LSMTree.Internal.RunFsPaths (
+
+activeDir :: SessionRoot -> FsPath
+activeDir (SessionRoot dir) = dir </> mkFsPath ["active"]
+
+snapshotsDir :: SessionRoot -> FsPath
+snapshotsDir (SessionRoot dir) = dir </> mkFsPath ["snapshots"]
+
+snapshot :: SessionRoot -> SnapshotName -> FsPath
+snapshot root (MkSnapshotName name) = snapshotsDir root </> mkFsPath [name]
+
+{-------------------------------------------------------------------------------
+  Snapshot name
+-------------------------------------------------------------------------------}
+
+newtype SnapshotName = MkSnapshotName FilePath
+  deriving (Eq, Ord)
+
+instance Show SnapshotName where
+  showsPrec d (MkSnapshotName p) = showsPrec d p
+
+-- | Create snapshot name.
+--
+-- The name may consist of lowercase characters, digits, dashes @-@ and underscores @_@.
+-- It must be non-empty and less than 65 characters long.
+-- It may not be a special filepath name.
+--
+-- >>> mkSnapshotName "main"
+-- Just "main"
+--
+-- >>> mkSnapshotName "temporary-123-test_"
+-- Just "temporary-123-test_"
+--
+-- >>> map mkSnapshotName ["UPPER", "dir/dot.exe", "..", "\\", "com1", "", replicate 100 'a']
+-- [Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing]
+--
+mkSnapshotName :: String -> Maybe SnapshotName
+mkSnapshotName s
+  | all isValid s
+  , len > 0
+  , len < 65
+  , System.FilePath.Posix.isValid s
+  , System.FilePath.Windows.isValid s
+  = Just (MkSnapshotName s)
+
+  | otherwise
+  = Nothing
+  where
+    len = length s
+    isValid c = ('a' <= c && c <= 'z') || ('0' <= c && c <= '9' ) || c `elem` "-_"
+
+{-------------------------------------------------------------------------------
+  Run paths
+-------------------------------------------------------------------------------}
+
 -- | The (relative) file path locations of all the files used by the run:
 --
--- Within the session root directory, the following files exist for active runs:
+-- The following files exist for a run:
 --
--- 1. @active/${n}.keyops@: the sorted run of key\/operation pairs
--- 2. @active/${n}.blobs@:  the blob values associated with the key\/operations
--- 3. @active/${n}.filter@: a Bloom filter of all the keys in the run
--- 4. @active/${n}.index@:  an index from keys to disk page numbers
--- 5. @active/${n}.checksums@: a file listing the crc32c checksums of the other
+-- 1. @${n}.keyops@: the sorted run of key\/operation pairs
+-- 2. @${n}.blobs@:  the blob values associated with the key\/operations
+-- 3. @${n}.filter@: a Bloom filter of all the keys in the run
+-- 4. @${n}.index@:  an index from keys to disk page numbers
+-- 5. @${n}.checksums@: a file listing the crc32c checksums of the other
 --    files
 --
 -- The representation doesn't store the full, name, just the number @n@. Use
 -- the accessor functions to get the actual names.
 --
-newtype RunFsPaths = RunFsPaths { runNumber :: Int }
-  deriving (Show, NFData)
+data RunFsPaths = RunFsPaths {
+    -- | The directory that run files live in.
+    runDir    :: !FsPath
```

Comment:
Alternatively, we could store the `SessionRoot` here. Otherwise, everyone constructing a `RunFsPaths` has to take care of adding the `active` subdirectory, which for example is not done in the tests and benchmarks.
Seems nicer to handle that concern (runs are always located in `${root}/active/${n}.keyops" etc.) in one central place, i.e. here.

Reply:
As a first version, I had the `SessionRoot` in there, but I think it's nicer if the `RunFsPaths` knows only about the directory it lives in. For most of our tests of the internals, it's not important that a run has a parent directory that is called `active` (which itself has a parent directory that is `${root}`). It inverts the responsibility: runs can now live anywhere by themselves, but it's the table/session that picks the `${root}/active` directory for the runs to live in, and that happens in a central place. There should be a single definition for how to create a run's paths in a session/table however, so what if I add this? Does that look reasonable?
 ```haskell
runPath :: SessionRoot -> Int -> RunFsPaths
runPath root = RunFsPaths (activeDir root)
```

> Otherwise, everyone constructing a RunFsPaths has to take care of adding the active subdirectory, which for example is not done in the tests and benchmarks.

So with this approach, none of our tests/benchmarks have to create additional directories. If the tests/benchmarks do pass in a `runDir` that does not exist, then our code will now fail early with an error telling us the directory does not exist (https://github.com/IntersectMBO/lsm-tree/pull/237/), which I think is also nicer because there is no silent directory creation that can cause errors down the line (which I experienced in #242)